src/Workspaces/Core/Portable/CodeActions/CodeAction.cs

```diff
@@ -259,33 +264,48 @@ internal Task<Document> GetChangedDocumentInternalAsync(CancellationToken cancel
+        protected Task<ImmutableArray<CodeActionOperation>> PostProcessAsync(IEnumerable<CodeActionOperation> operations, CancellationToken cancellationToken)
+            => PostProcessAsync(originalSolution: null!, operations, cancellationToken);
+
+        internal async Task<ImmutableArray<CodeActionOperation>> PostProcessAsync(
+            Solution originalSolution, IEnumerable<CodeActionOperation> operations, CancellationToken cancellationToken)
         {
-            var arrayBuilder = new ArrayBuilder<CodeActionOperation>();
+            using var result = TemporaryArray<CodeActionOperation>.Empty;
 
             foreach (var op in operations)
             {
                 if (op is ApplyChangesOperation ac)
                 {
-                    arrayBuilder.Add(new ApplyChangesOperation(await this.PostProcessChangesAsync(ac.ChangedSolution, cancellationToken).ConfigureAwait(false)));
+                    result.Add(new ApplyChangesOperation(await this.PostProcessChangesAsync(originalSolution, ac.ChangedSolution, cancellationToken).ConfigureAwait(false)));
                 }
                 else
                 {
-                    arrayBuilder.Add(op);
+                    result.Add(op);
                 }
             }
 
-            return arrayBuilder.ToImmutableAndFree();
+            return result.ToImmutableAndClear();
         }
 
         /// <summary>
-        ///  Apply post processing steps to solution changes, like formatting and simplification.
+        /// Apply post processing steps to solution changes, like formatting and simplification.
         /// </summary>
         /// <param name="changedSolution">The solution changed by the <see cref="CodeAction"/>.</param>
         /// <param name="cancellationToken">A cancellation token</param>
-        protected async Task<Solution> PostProcessChangesAsync(Solution changedSolution, CancellationToken cancellationToken)
+        protected Task<Solution> PostProcessChangesAsync(Solution changedSolution, CancellationToken cancellationToken)
+            => PostProcessChangesAsync(originalSolution: null!, changedSolution, cancellationToken);
+
+        internal async Task<Solution> PostProcessChangesAsync(
+            Solution originalSolution,
+            Solution changedSolution,
+            CancellationToken cancellationToken)
         {
-            var solutionChanges = changedSolution.GetChanges(changedSolution.Workspace.CurrentSolution);
+            // originalSolution is only null on backward compatible codepaths.  In that case, we get the workspace's
+            // current solution.  This is not ideal (as that is a mutable field that could be changing out from
+            // underneath us).  But it's the only option we have for the compat case with existing public extension
+            // points.
+            originalSolution ??= changedSolution.Workspace.CurrentSolution;
```

Comment:
💭 I was going to say add a debug assertion here, but I'm pretty sure we also pass null through this path in the analyzer testing library (which is fine and comes through a public compat API).

Reply:
yup :)